### PR TITLE
Refine effects based on optimizer-derived information

### PR DESCRIPTION
**Pull request by Keno**
The optimizer may be able to derive information that is not available to inference. For example, it may SROA a mutable value to derive additional constant information. Additionally, some effects, like :consistent are path-dependent and should ideally be scanned once all optimizations are done. Now, there is a bit of a complication that we have generally so far taken the position that the optimizer may do non-IPO-safe optimizations, although in practice we never actually implemented any. This was a sensible choice, because we weren't really doing anything with the post-optimized IR other than feeding it into codegen anyway. However, with irinterp and this change, there's now two consumers of IPO-safely optimized IR. I do still think we may at some point want to run passes that allow IPO-unsafe optimizations, but we can always add them at the end of the pipeline.

With these changes, the effect analysis is a lot more precise. For example, we can now derive :consistent for these functions:
```
function f1(b)
    if Base.inferencebarrier(b)
        error()
    end
    return b
end

function f3(x)
    @fastmath sqrt(x)
    return x
end
```
and we can derive `:nothrow` for this function:
```
function f2()
    if Ref(false)[]
        error()
    end
    return true
end
```

### CODE REVIEW

**Review by vtjnash — DISMISSED** [review]
> Now, there is a bit of a complication that we have generally so far taken the position that the optimizer may do non-IPO-safe optimizations, although in practice we never actually implemented any. This was a sensible choice, because we weren't really doing anything with the post-optimized IR other than feeding it into codegen anyway. However, with irinterp and this change, there's now two consumers of IPO-safely optimized IR. I do still think we may at some point want to run passes that allow IPO-unsafe optimizations, but we can always add them at the end of the pipeline.

We don't do as many as we should, but there is a long way between doing only a few and being prohibited from doing any ever. For a direct example, `Base.inferencebarrier` is required to be optimized away, but must not improve IPO information afterwards. Other similar examples may include 1. the legality of removing boundscheck depends on having determined whether it is guaranteed to be not inlined (which premise is broken by this PR). 2. folding Expr(:new) with undef fields to a constant and other similar operations with undefined bits such as accessing `.parameters` or doing math with floating point numbers assumes IPO cannot see that change to the implementation 3. DCE cannot mark a value as unused after all loads are removed (changing it to readnone), since later IPO effects may incorrectly assume this argument could be optimized away in the caller

**Comment by Keno** [comment]
> We don't do as many as we should, but there is a long way between doing only a few and being prohibited from doing any ever. 

You're not prohibited from doing any ever, you're just prohibited from doing them before the analysis pass runs. There's no problem doing more things afterwards.

> For a direct example, `Base.inferencebarrier` is required to be optimized away, but must not improve IPO information afterwards.

Yes, that needed to be adjusted in this PR.

> Other similar examples may include 1. the legality of removing boundscheck depends on having determined whether it is guaranteed to be not inlined (which premise is broken by this PR). 

This PR is on the path of overhauling this mechanism. It's possible there's some extra tweaking that needs to happen here in this intermediate step.

> folding Expr(:new) with undef fields to a constant and other similar operations with undefined bits such as accessing `.parameters` or doing math with floating point numbers assumes IPO cannot see that change to the implementation 

I don't believe we do this in the current pass pipeline.

> DCE cannot mark a value as unused after all loads are removed (changing it to readnone), since later IPO effects may incorrectly assume this argument could be optimized away in the caller

We don't currently derive such information interprocedurally.

**Comment by vtjnash** [comment]
> you're just prohibited from doing them before the analysis pass runs

In particular though, this means the analysis pass must run before inlining, since inlining already ran all passes. Which seems hardly indistinguishable from not being able to do any useful optimizations?

> This PR is on the path of overhauling this mechanism

I am talking about the default environment, which is not affected by that PR. In the default cases where it is not forced on or forced off, the effect of inlining on boundscheck branches is known to not be consistent, but that inconsistency is usually removed by the inlining pass. Though I suppose that falls into the larger bucket above that inlining is not allowed before the refinement analysis, so it might not matter as an independent point.

**Comment by Keno** [comment]
As discussed today, we'll stop doing :boundscheck transforms in inlining and move those to codegen instead. Then we can cache the IR as IPO safe. At whatever point in the future we want to do IPO unsafe IR transforms at the julia level, we'll have a callback from codegen back into julia that does that.

**Comment by vtjnash** [comment]
How did we decide to deal with the fact this means that the result of codegen (e.g. validity of using concrete eval) will no longer be consistent with the IPO flags on the same CodeInfo object?

**Comment by Keno** [comment]
> How did we decide to deal with the fact this means that the result of codegen (e.g. validity of using concrete eval) will no longer be consistent with the IPO flags on the same CodeInfo object?

You mean CodeInstance? They already have two sets of flags, but if something is determined for IPO, it should always be valid even after codegen transforms.

File: base/compiler/optimize.jl [hunk 1]
```diff
@@ -509,7 +711,7 @@ function run_passes(
     @pass "compact 2" ir = compact!(ir)
```
Comment (vtjnash) [inline base/compiler/optimize.jl:711]: This pass currently seems to not be IPO safe, since the access to `ci.propagate_inbounds` (and `:boundscheck` removal folding) is not currently IPO safe

Comment (Keno) [reply base/compiler/optimize.jl:711]: Yes, I'm working on that right now.

File: base/compiler/abstractinterpretation.jl [hunk 2]
```diff
@@ -2529,6 +2494,20 @@ function abstract_eval_statement_expr(interp::AbstractInterpreter, e::Expr, vtyp
...
+        condval = maybe_extract_const_bool(condt)
+        t = Nothing
+        effects = EFFECTS_THROWS
+        if condval isa Bool
+            if condval
+                effects = EFFECTS_TOTAL
+            else
+                t = Union{}
+            end
+        elseif !hasintersect(condt, Bool)
```
Comment (aviatesk) [inline base/compiler/abstractinterpretation.jl:2508]:
```suggestion
        elseif !hasintersect(widenconst(condt), Bool)
```

File: base/compiler/effects.jl [hunk 3]
```diff
@@ -139,10 +136,13 @@ const EFFECT_FREE_IF_INACCESSIBLEMEMONLY = 0x01 << 1
...
-const EFFECTS_TOTAL    = Effects(ALWAYS_TRUE,  ALWAYS_TRUE,  true,  true,  true,  ALWAYS_TRUE,  true,  true,  true)
-const EFFECTS_THROWS   = Effects(ALWAYS_TRUE,  ALWAYS_TRUE,  false, true,  true,  ALWAYS_TRUE,  true,  true,  true)
-const EFFECTS_UNKNOWN  = Effects(ALWAYS_FALSE, ALWAYS_FALSE, false, false, false, ALWAYS_FALSE, false, true,  true) # unknown mostly, but it's not overlayed and noinbounds at least (e.g. it's not a call)
-const _EFFECTS_UNKNOWN = Effects(ALWAYS_FALSE, ALWAYS_FALSE, false, false, false, ALWAYS_FALSE, false, false, false) # unknown really
+# :noub bits
+const NOUB_IF_NOINBOUNDS = 0x01 << 1
+
+const EFFECTS_TOTAL    = Effects(ALWAYS_TRUE,  ALWAYS_TRUE,  true,  true,  true,  ALWAYS_TRUE,  ALWAYS_TRUE,  true)
+const EFFECTS_THROWS   = Effects(ALWAYS_TRUE,  ALWAYS_TRUE,  false, true,  true,  ALWAYS_TRUE,  ALWAYS_TRUE,  true)
+const EFFECTS_UNKNOWN  = Effects(ALWAYS_FALSE, ALWAYS_FALSE, false, false, false, ALWAYS_FALSE, ALWAYS_FALSE, true) # unknown mostly, but it's not overlayed and noinbounds at least (e.g. it's not a call)
```
Comment (aviatesk) [inline base/compiler/effects.jl:144]:
```suggestion
const EFFECTS_UNKNOWN  = Effects(ALWAYS_FALSE, ALWAYS_FALSE, false, false, false, ALWAYS_FALSE, ALWAYS_FALSE, true) # unknown mostly, but it's not overlayed at least (e.g. it's not a call)
```

File: base/compiler/effects.jl [hunk 4]
```diff
@@ -193,18 +190,18 @@ is_nothrow(effects::Effects)             = effects.nothrow
 is_terminates(effects::Effects)          = effects.terminates
 is_notaskstate(effects::Effects)         = effects.notaskstate
 is_inaccessiblememonly(effects::Effects) = effects.inaccessiblememonly === ALWAYS_TRUE
-is_noub(effects::Effects)                = effects.noub
+is_noub(effects::Effects, noinbounds=true) = effects.noub === ALWAYS_TRUE || (noinbounds && effects.noub === NOUB_IF_NOINBOUNDS)
```
Comment (aviatesk) [inline base/compiler/effects.jl:193]:
```suggestion
is_noub(effects::Effects, noinbounds::Bool=true) = effects.noub === ALWAYS_TRUE || (noinbounds && effects.noub === NOUB_IF_NOINBOUNDS)
```

File: base/compiler/effects.jl [hunk 5]
```diff
@@ -193,18 +190,18 @@ is_nothrow(effects::Effects)             = effects.nothrow
...
 is_nonoverlayed(effects::Effects)        = effects.nonoverlayed
 
 # implies `is_notaskstate` & `is_inaccessiblememonly`, but not explicitly checked here
-is_foldable(effects::Effects) =
+is_foldable(effects::Effects, noinbounds=true) =
```
Comment (aviatesk) [inline base/compiler/effects.jl:197]:
```suggestion
is_foldable(effects::Effects, noinbounds::Bool=true) =
```

File: base/compiler/effects.jl [hunk 6]
```diff
@@ -193,18 +190,18 @@ is_nothrow(effects::Effects)             = effects.nothrow
...
     is_consistent(effects) &&
-    is_noub(effects) &&
+    is_noub(effects, noinbounds) &&
     is_effect_free(effects) &&
     is_terminates(effects)
 
-is_foldable_nothrow(effects::Effects) =
-    is_foldable(effects) &&
+is_foldable_nothrow(effects::Effects, noinbounds=true) =
```
Comment (aviatesk) [inline base/compiler/effects.jl:203]:
```suggestion
is_foldable_nothrow(effects::Effects, noinbounds::Bool=true) =
```

File: base/compiler/optimize.jl [hunk 7]
```diff
@@ -15,27 +15,29 @@ const SLOT_USEDUNDEF    = 32 # slot has uses that might raise UndefVarError
 
 # NOTE make sure to sync the flag definitions below with julia.h and `jl_code_info_set_ir` in method.c
 
-const IR_FLAG_NULL        = 0x00
+const IR_FLAG_NULL        = UInt32(1)
```
Comment (aviatesk) [inline base/compiler/optimize.jl:18]:
```suggestion
const IR_FLAG_NULL        = UInt32(0)
```

File: base/reflection.jl [hunk 8]
```diff
@@ -1647,9 +1647,9 @@ function infer_effects(@nospecialize(f), @nospecialize(types=default_tt(f));
     for match in matches.matches
         match = match::Core.MethodMatch
         frame = Core.Compiler.typeinf_frame(interp,
-            match.method, match.spec_types, match.sparams, #=run_optimizer=#false)
+            match.method, match.spec_types, match.sparams, #=run_optimizer=#true)
```
Comment (aviatesk) [inline base/reflection.jl:1650]: New optional argument `post_optimization_analysis::Bool=true` might be useful for testing/debugging purposes. We can add it later when needed though.

File: base/compiler/abstractinterpretation.jl [hunk 9]
```diff
@@ -853,14 +847,8 @@ function concrete_eval_eligible(interp::AbstractInterpreter,
...
     end
-    if !effects.noinbounds && stmt_taints_inbounds_consistency(sv)
-        # If the current statement is @inbounds or we propagate inbounds,
-        # the call's :consistent-cy is tainted and not consteval eligible.
-        add_remark!(interp, sv, "[constprop] Concrete evel disabled for inbounds")
-        return :none
-    end
     mi = result.edge
-    if mi !== nothing && is_foldable(effects)
+    if mi !== nothing && is_foldable(effects, !stmt_taints_inbounds_consistency(sv))
```
Comment (aviatesk) [inline base/compiler/abstractinterpretation.jl:851]: Maybe we need a similar change to `concrete_eval_invoke`?

File: base/compiler/optimize.jl [hunk 10]
```diff
@@ -453,9 +457,262 @@ function finish(interp::AbstractInterpreter, opt::OptimizationState,
...
+    any_conditional_ub = false
+    had_trycatch = false
+
+    scanner = BBScanner(ir)
+
+    effects = result.ipo_effects
+
+    cfg = nothing
+    domtree = nothing
+    function get_augmented_domtree()
```
Comment (aviatesk) [inline base/compiler/optimize.jl:519]: Probably better to set up a new struct 
```julia
struct AugumentedDomtree
    cfg::CFG
    domtree::DomTree
end
```
and make `get_augumented_domtree` return the struct instead for better inferrability.

File: base/compiler/optimize.jl [hunk 11]
```diff
@@ -453,9 +457,262 @@ function finish(interp::AbstractInterpreter, opt::OptimizationState,
...
+    end
+
+    function is_getfield_boundscheck(inst::Instruction)
+        stmt = inst[:stmt]
+        is_known_call(stmt, getfield, ir) || return false
+        length(stmt.args) < 4 && return false
+        boundscheck = stmt.args[end]
+        argextype(boundscheck, ir) === Bool || return false
+        isa(boundscheck, SSAValue) || return false
+        return true
```
Comment (aviatesk) [inline base/compiler/optimize.jl:545]: Include `isexpr(ir[boundscheck][:stmt], :boundscheck)` here?

File: base/compiler/optimize.jl [hunk 12]
```diff
@@ -453,9 +457,262 @@ function finish(interp::AbstractInterpreter, opt::OptimizationState,
...
+                            any_non_boundscheck_inconsistent && break
+                        end
+                    end
+                    any_non_boundscheck_inconsistent || continue
+                elseif isa(stmt, ReturnNode)
+                    all_retpaths_consistent = false
+                else isa(stmt, GotoIfNot)
+                    bb = block_for_inst(ir, idx)
+                    sa, sb = ir.cfg.blocks[bb].succs
+                    for succ in iterated_dominance_frontier(ir.cfg, BlockLiveness((sa, sb), nothing), get!(lazydomtree))
```
Comment (aviatesk) [inline base/compiler/optimize.jl:682]:
```suggestion
                    for succ in iterated_dominance_frontier(ir.cfg, BlockLiveness(ir.cfg.blocks[bb].succs, nothing), get!(lazydomtree))
```
and then we can keep typing `BlockLiveness.def_bbs::Vector{Int}`.

File: base/compiler/ssair/ir.jl [hunk 13]
```diff
@@ -1390,7 +1410,7 @@ function process_node!(compact::IncrementalCompact, result_idx::Int, inst::Instr
         if cfg_transforms_enabled
             # Rename phi node edges
             let bb_rename_pred=bb_rename_pred
-                map!(i::Int32->bb_rename_pred[i], stmt.edges, stmt.edges)
+                map!(i::Int32->i == 0 ? 0 : bb_rename_pred[i], stmt.edges, stmt.edges)
```
Comment (aviatesk) [inline base/compiler/ssair/ir.jl:1413]: Is this change necessary for this PR, or a fix for some related issues?

File: base/compiler/optimize.jl [hunk 14]
```diff
@@ -453,9 +457,262 @@ function finish(interp::AbstractInterpreter, opt::OptimizationState,
...
+        all_effect_free &= (flag & IR_FLAG_EFFECT_FREE) != 0
+        all_nothrow &= (flag & IR_FLAG_NOTHROW) != 0
+        if (flag & IR_FLAG_NOUB) == 0
+            if !is_conditional_noub(inst)
+                all_noub = false
+            end
+        end
+    end
+
+    function scan_inconsistency!(inst::Instruction, idx)
```
Comment (aviatesk) [inline base/compiler/optimize.jl:570]:
```suggestion
    function scan_inconsistency!(inst::Instruction, idx::Int)
```

File: base/compiler/optimize.jl [hunk 15]
```diff
@@ -453,9 +457,262 @@ function finish(interp::AbstractInterpreter, opt::OptimizationState,
...
+                    stmt_inconsistent |= val.id in inconsistent
+                    count!(tpdum, val)
+                end
+            end
+        end
+        stmt_inconsistent && push!(inconsistent, idx)
+        return stmt_inconsistent
+    end
+
+    function scan_stmt!(inst, idx, lstmt, bb)
```
Comment (aviatesk) [inline base/compiler/optimize.jl:596]:
```suggestion
    function scan_stmt!(inst::Instruction, idx::Int, lstmt::Int, bb::Int)
```

File: base/compiler/optimize.jl [hunk 16]
```diff
@@ -453,9 +457,262 @@ function finish(interp::AbstractInterpreter, opt::OptimizationState,
...
+        stmt = inst[:inst]
+        flag = inst[:flag]
+
+        if isexpr(stmt, :enter)
+            # try/catch not yet modeled
+            had_trycatch = true
+            return true
```
Comment (aviatesk) [inline base/compiler/optimize.jl:603]:
```suggestion
            return false
```
Shouldn't return `false` to bail out the scan?

File: base/compiler/optimize.jl [hunk 17]
```diff
@@ -453,9 +457,262 @@ function finish(interp::AbstractInterpreter, opt::OptimizationState,
...
+                end
+            end
+        end
+
+        return true
+    end
+    if !scan!(scan_stmt!, scanner, true)
+        if !all_retpaths_consistent
+            # No longer any dataflow concerns, just scan the flags
+            scan!(scanner, false) do inst, idx, lstmt, bb
```
Comment (aviatesk) [inline base/compiler/optimize.jl:645]:
```suggestion
            scan!(scanner, false) do inst::Instruction, idx::Int, lstmt::Int, bb::Int
```

File: base/compiler/optimize.jl [hunk 18]
```diff
@@ -453,9 +457,262 @@ function finish(interp::AbstractInterpreter, opt::OptimizationState,
...
+function visit_bb_phis!(callback, ir::IRCode, bb::Int)
+    stmts = ir.cfg.blocks[bb].stmts
+    for idx in stmts
+        stmt = ir[SSAValue(idx)][:inst]
+        if !isa(stmt, PhiNode)
+            if !is_valid_phiblock_stmt(stmt)
+                return
+            end
+        else
+            callback(idx, stmt)
```
Comment (aviatesk) [inline base/compiler/optimize.jl:469]:
```suggestion
            callback(idx)
```
`stmt::PhiNode` seems to be never used.

**Comment by aviatesk** [comment]
> I was seeing a fair number of inference instabilities on master
> and tracked this down to us deleting the inferred code after
> codegen making it unavailable to irinterp.

Doesn't absint-based constprop' usually lead to better inference? I'm in favor of preserving sources for irinterp, but I want to assert my understanding that irinterp is faster but less precise than the absint constprop'.

**Comment by Keno** [comment]
> Doesn't absint-based constprop' usually lead to better inference? I'm in favor of preserving sources for irinterp, but I want to assert my understanding that irinterp is faster but less precise than the absint constprop'.

I don't think it's necessarily always true that irinterp is less precise than absint. For example, irinterp can in theory propagate through more complicated sroa patterns. I didn't track down exactly what was going on in this case, but the failing test was arrayops 1746.

File: base/compiler/abstractinterpretation.jl (hunk 9 above)
Comment (Keno) [reply base/compiler/abstractinterpretation.jl:851]: Actually, I don't think it's required here at all. Both concrete eval and semi concrete eval are safe here independent of the state of the inbounds flag.

File: base/compiler/optimize.jl (hunk 11 above)
Comment (Keno) [reply base/compiler/optimize.jl:545]: This is meant to check for having a boundscheck argument rather than the boundscheck expr, I'll rename the function.

File: base/compiler/ssair/ir.jl (hunk 13 above)
Comment (Keno) [reply base/compiler/ssair/ir.jl:1413]: It fixed some test failure, but I don't think it's related to this PR other than by coincidence. There's a special `i == 0` predecessor for entry into catch blocks.

**Comment by Keno** [comment]
@nanosoldier `runtests()`

**Comment by Keno** [comment]
MRE for the Plots issue seen on PkgEval:
```
julia> Plots._cbar_unique([nothing], "fill alpha")
Internal error: encountered unexpected error in runtime:
BoundsError(a=Array{Int64, (87,)}[1, 1, 0, 1, 1, 2, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 3, 0, 1, 1, 2, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0], i=(94,))
ijl_bounds_error_ints at /home/keno/julia/src/rtutils.c:201
getindex at ./essentials.jl:13 [inlined]
#346 at ./compiler/ssair/ir.jl:1723 [inlined]
foreachssa at ./compiler/utilities.jl:373
maybe_erase_unused! at ./compiler/ssair/ir.jl:1722
```

**Comment by Keno** [comment]
Non-Plots reproducer:
```
@inline function bar()
	x = 1
	try
		x = 2
		if Ref{Bool}(false)[]
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			Base.donotdelete(x)
			x = 3
		end
	catch
		return x
	end
end

function foo()
	if Ref{Bool}(false)[]
		bar()
	end
	return 4
end
```

**Comment by Keno** [comment]
@nanosoldier `runtests(["Sherlock", "NonconvexIpopt", "OptimizationNLopt", "Test", "InferOpt", "NonconvexNOMAD", "OptimizationOptimisers", "GaussianKDEs", "Reel", "StochasticAD", "WaterWaves1D", "ChargeTransport", "NormalizingFlows", "EnKF", "NonconvexJuniper", "BAT", "NeuralQuantumState", "Robotlib", "ReferenceTests", "PerformanceProfilingHttpEndpoints", "OptimalTransmissionRouting", "ClimaCorePlots", "TexasHoldem", "FluxOptTools", "DecisionMakingPolicies", "NonconvexCore", "NonconvexPavito", "NonconvexMMA", "ConditionalDists", "Pseudospectra", "FourierSpaces", "ProtoBuf", "ReinforcementLearningCore", "RemoteFiles", "Diagonalizations", "SurrogatesAbstractGPs", "RRTMGP", "DifferentiableFlatten", "KernelAbstractions", "ProbabilityBoundsAnalysis", "PlantBiophysics", "SurrogatesMOE", "MIRTjim", "Integrals", "UltraDark", "MLJGaussianProcesses", "BaytesMCMC", "Boltz", "OptimizationMOI", "Plots", "RetentionParameterEstimator", "MIRT", "ObjectOriented", "BaytesDiff", "WaveguideQED", "EarthSciMLBase", "Sunny", "RegularizedProblems", "ReactionSensitivity", "NonconvexPercival", "ConScape", "ImplicitDifferentiation", "SBMLToolkit", "OptimizationMetaheuristics", "OptimizationOptimJL", "BM3DDenoise", "ParameterHandling", "ModelWrappers", "RAPIDS", "GeoEstimation", "ReliabilityOptimization", "NonconvexNLopt", "BigRiverQTLPlots", "Anatta", "MultiscaleGraphSignalTransforms", "GeoStatsPlots", "SurrogatesFlux", "ProfileEndpoints", "PtFEM", "LowLevelParticleFilters", "CoDa", "SyntacticModels", "NonconvexUtils", "ADRIA"])`

**Comment by Keno** [comment]
@nanosoldier `runbenchmarks("inference", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/50805#issuecomment-1698415764) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/e9b4026_vs_a84882e/report.md).

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/50805#issuecomment-1698415196) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/c84f526_vs_8ace9c5/report.html) is available.

**Comment by Keno** [comment]
> [Your benchmark job](https://github.com/JuliaLang/julia/pull/50805#issuecomment-1698415764) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/e9b4026_vs_a84882e/report.md).

I looked into this - the couple percent in general are fine. This does more work, so that's not unexpected. I looked into the large regression in `["inference", "abstract interpretation", "many_const_calls"]` - that turns out to just be because the benchmark forcibly disables the optimizer and this PR moves the hack that we used to have in inference to prove consistency of a literal `getfield(x, f, Expr(:boundscheck))` into the dataflow pass instead. Since this is my benchmark, I'm just gonna go ahead and say this is fine. `may_optimize() = false` is not really an optimized configuration (it already messes with precision because it changes the inlining cost model). We should revisit whether it makes sense as a benchmark. 

I'm gonna go ahead and merge this, I have a few more things that can make this better, but I'll save them for later before this PR gets even bigger than it already is.
